esx_service/utils/vm_listener.py

```diff
@@ -0,0 +1,154 @@
+    """
+    si = vmdk_ops.get_si()
+
+    while not si:
+        # If hostd is not up yet, sleep for a while and try again
+        # Proceed further only after you get si instance
+        logging.warn("VMChangeListener couldn't connect to hostd. Retrying after %s seconds", HOSTD_RECONNECT_INTERVAL)
+        time.sleep(HOSTD_RECONNECT_INTERVAL)
+        si = vmdk_ops.get_si()
+
+    pc = si.content.propertyCollector
+    err_msg = create_vm_powerstate_filter(pc, si.content.rootFolder)
+    if err_msg:
+        return err_msg
+    # listen to changes
+    listen_vm_propertychange(pc)
+
+
+def create_vm_powerstate_filter(pc, from_node):
+    """
+    Create a filter spec to list to VM power state changes
+    """
+
+    filterSpec = vmodl.query.PropertyCollector.FilterSpec()
+    objSpec = vmodl.query.PropertyCollector.ObjectSpec(obj=from_node,
+                                                       selectSet=vm_folder_traversal())
+    filterSpec.objectSet.append(objSpec)
+    # Add the property specs
+    propSpec = vmodl.query.PropertyCollector.PropertySpec(type=vim.VirtualMachine, all=False)
+    propSpec.pathSet.append(VM_POWERSTATE)
+    filterSpec.propSet.append(propSpec)
+    try:
+        pcFilter = pc.CreateFilter(filterSpec, True)
+        atexit.register(pcFilter.Destroy)
+        return None
+    except Exception as e:
+        err_msg = "Problem creating PropertyCollector filter: {}".format(str(e))
+        logging.error(err_msg)
+        return err_msg
+
+
+def listen_vm_propertychange(pc):
+    """
+    Waits for updates on powerstate of VMs. If powerstate is poweroff,
+    detach the dvs managed volumes attached to VM
+    """
+    logging.info("VMChangeListener thread started")
+    version = ''
+    while True:
+        result = pc.WaitForUpdates(version)
+        try:
+            # process the updates result
+            for filterSet in result.filterSet:
+                for objectSet in filterSet.objectSet:
+                    if objectSet.kind != 'modify':
+                        continue
+                    for change in objectSet.changeSet:
+                        # if the event was powerOff for a VM, set the status of all
+                        # docker volumes attached to the VM to be detached
+                        if change.name != VM_POWERSTATE or change.val != POWERSTATE_POWEROFF:
+                            continue
+
+                        moref = getattr(objectSet, 'obj', None)
+                        # Do we need to alert the admin? how?
+                        if not moref:
+                            logging.error("Could not retrieve the VM managed object.")
+                            continue
+
+                        logging.info("VM poweroff change found for %s", moref.config.name)
+
+                        set_device_detached(moref)
+        except Exception as e:
+            # Do we need to alert the admin? how?
+            logging.error("VMChangeListener: error %s", str(e))
+
```

Comment:
Can include a test to restart hostd when this thread is waiting. Does the thread continue to get updates once hostd restarts?

Other than this no comments on the code. Only check behavior if hostd restart when this thread is waiting updates.